### PR TITLE
Tests connection

### PR DESCRIPTION
**Pull request by m00lecule**
Reason of this PR is to increase tests code coverage of Janus.Connection module.

I have introduced method of thread safe mocking on Janus.DateTimeUtils using Mox.

### CODE REVIEW

File: lib/connection.ex [hunk 1]
```diff
@@ -198,11 +203,14 @@ defmodule Janus.Connection do
...
+  def handle_info(
+        :cleanup,
+        state(pending_calls_table: pending_calls_table, cleanup_interval: cleanup_interval) = s
+      ) do
     Transaction.cleanup_old_transactions(pending_calls_table)
 
-    Process.send_after(self(), :cleanup, @cleanup_interval)
-    {:noreply, state}
+    Process.send_after(self(), :cleanup, cleanup_interval)
+    {:noreply, s}
```
Comment (Hajto) [inline lib/connection.ex]: Use full state name. There is little benefit of using s shorthand.

File: test/support/module/fake_transport.ex [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  def default_session_id(), do: @session_id
+
+  @impl true
+  def connect({id}) do
+    {:ok, %{message_receiver: self(), id: id}}
+  end
+
+  def send(message, _timeout, %{message_receiver: receiver} = state) do
+    payload = Jason.encode!(message) |> Jason.decode!()
+    :ok = handle_payload(payload, receiver)
```
Comment (Hajto) [inline test/support/module/fake_transport.ex]: I can't grasp why you are doing this here. `handle_payload` is a function that you control and you manually return `:ok` from every handle. Why don't we just ignore the return if it is a purely side effect focused function?

File: test/support/module/fake_transport.ex [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+         receiver
+       ) do
+    send(receiver, %{"janus" => "ack", "transaction" => transaction})
+    :ok
+  end
+
+  defp handle_payload(%{"transaction" => transaction} = msg, receiver) do
+    send(receiver, %{"janus" => "success", "data" => msg, "transaction" => transaction})
+    :ok
+  end
```
Comment (Hajto) [inline test/support/module/fake_transport.ex]:
```suggestion
  defp handle_payload(%{"janus" => "create", "transaction" => transaction}, receiver) do
    send(receiver, %{
      "janus" => "success",
      "transaction" => transaction,
      "data" => %{"id" => @session_id}
    })
  end

  defp handle_payload(
         %{"janus" => "keepalive", "session_id" => _session_id, "transaction" => transaction},
         receiver
       ) do
    send(receiver, %{"janus" => "ack", "transaction" => transaction})
  end

  defp handle_payload(%{"transaction" => transaction} = msg, receiver) do
    send(receiver, %{"janus" => "success", "data" => msg, "transaction" => transaction})
  end
```

File: test/support/module/fake_transport.ex (hunk 2 above)
Comment (Hajto) [inline test/support/module/fake_transport.ex]:
```suggestion
    handle_payload(payload, receiver)
```

File: test/support/module/fake_transport.ex [hunk 4]
```diff
@@ -0,0 +1,81 @@
+defmodule Janus.Support.FakeTransport do
```
Comment (Hajto) [inline test/support/module/fake_transport.ex]: Take a look at current master, I think there has been added MockTransport that is basically this module but generalized. A lot of tests has been rewritten using that module. Take a look at it, and make sure you are not doubling the work done, ok?

Comment (Hajto) [reply test/support/module/fake_transport.ex]: Here is the link:
https://github.com/software-mansion-labs/elixir-janus/blob/master/lib/mock_transport.ex

Comment (Hajto) [reply test/support/module/fake_transport.ex]: Looking at both these modules they seem to be used in a very very similar way, but the one that @Qizot wrote is general purpose.

File: config/config.exs [hunk 5]
```diff
@@ -0,0 +1,3 @@
+use Mix.Config
```
Comment (Hajto) [inline config/config.exs]: This has been deprecated.
```
A new Config module has been added to Elixir. The previous configuration API, Mix.Config,
 was part of the Mix build tool. But since releases provide runtime configuration and Mix is
not included in releases, we ported the Mix.Config API to Elixir. In other words,
use Mix.Config has been soft-deprecated in favor of import Config.
```
Source: https://github.com/elixir-lang/elixir/releases/tag/v1.9.0

File: config/dev.exs [hunk 6]
```diff
@@ -0,0 +1 @@
+use Mix.Config
```
Comment (Hajto) [inline config/dev.exs]: This has been deprecated.
```
A new Config module has been added to Elixir. The previous configuration API, Mix.Config,
 was part of the Mix build tool. But since releases provide runtime configuration and Mix is
not included in releases, we ported the Mix.Config API to Elixir. In other words,
use Mix.Config has been soft-deprecated in favor of import Config.
```
Source: https://github.com/elixir-lang/elixir/releases/tag/v1.9.0

File: config/test.exs [hunk 7]
```diff
@@ -0,0 +1,3 @@
+use Mix.Config
```
Comment (Hajto) [inline config/test.exs]: This has been deprecated.
```
A new Config module has been added to Elixir. The previous configuration API, Mix.Config,
 was part of the Mix build tool. But since releases provide runtime configuration and Mix is
not included in releases, we ported the Mix.Config API to Elixir. In other words,
use Mix.Config has been soft-deprecated in favor of import Config.
```
Source: https://github.com/elixir-lang/elixir/releases/tag/v1.9.0

File: lib/connection.ex [hunk 8]
```diff
@@ -128,7 +129,10 @@ defmodule Janus.Connection do
   # Callbacks
 
   @impl true
-  def init({transport_module, transport_args, handler_module, handler_args}) do
+  def init(
+        {transport_module, transport_args, handler_module, handler_args},
+        cleanup_interval \\ @cleanup_interval
```
Comment (Hajto) [inline lib/connection.ex]: 1. Why this change needed to happen?
2. If that change is going to stay I would either
a) change the name
```suggestion
        cleanup_interval \\ @default_cleanup_interval
```
Just because it makes much more sense semantically.
b) put that thing into config and read that in init. The reason being if this is going to be changed I guess it's going to be done through configuration, and not through a code change.

File: test/connection_test.exs [hunk 9]
```diff
@@ -11,4 +41,164 @@ defmodule Janus.ConnectionTest do
...
+
+  describe "call should" do
+    defmodule ValidTransportMock do
+      def send(_, _, _), do: {}
+    end
+
+    test "send transaction through transport and add transaction to transaction store", %{
+      table: table
+    } do
+      with_mock(ValidTransportMock, send: fn _payload, _timeout, _state -> {:ok, "dummy"} end) do
```
Comment (Hajto) [inline test/connection_test.exs:165]: this is an add hoc mock and module has async: true

Comment (Hajto) [reply test/connection_test.exs:165]: If you use Transport from master this mock will not be needed I think. I might be wrong though.

File: test/connection_test.exs [hunk 10]
```diff
@@ -1,6 +1,36 @@
...
+    zone_abbr: "CET",
+    hour: 23,
+    minute: 0,
+    second: 7,
+    utc_offset: 3600,
+    std_offset: 0,
+    time_zone: "Etc/UTC"
+  }
+
+  setup do
```
Comment (Hajto) [inline test/connection_test.exs:30]: A lot of tests here uses state. You can just make a state that fits most of those cases and put it into setup_all clause. It's executed once per file, and its result is found in any test.

File: test/connection_test.exs [hunk 11]
```diff
@@ -1,6 +1,36 @@
 defmodule Janus.ConnectionTest do
```
Comment (Hajto) [inline test/connection_test.exs:1]: Try looking at the MockTransport and we shall revisit this, ok?

File: test/session_test.exs [hunk 12]
```diff
@@ -38,9 +38,13 @@ defmodule Janus.SessionTest do
...
+  setup :set_mox_global
+
   setup do
+    stub(DateTimeMock, :utc_now, &DateTime.utc_now/0)
+
     {:ok, connection} =
-      Connection.start(Janus.MockTransport, @request_response_pairs, FakeHandler, {})
+      Connection.start(Janus.MockTransport, @request_response_pairs, FakeHandler, [], [])
 
     %{connection: connection}
```
Comment (Hajto) [inline test/session_test.exs:51]: setup on exit here
https://hexdocs.pm/ex_unit/ExUnit.Callbacks.html#on_exit/2
It should work.

File: test/support/macro/connection_helpers.ex [hunk 13]
```diff
@@ -1,11 +1,12 @@
-defmodule Janus.HandlerTest.CallbackHelper do
-  # helper module that declares macro for testing `Janus.Handler`'s callback calls
-
-  # simple transport returning message given to `handle_info/2` without changes
-  # used by `Janus.Connection.handle_info/2`
-  defmodule SimpleTransport do
-    def handle_info(message, state) do
-      {:ok, message, state}
+defmodule Janus.Support.Macro do
+  defmacro assert_next_receive(pattern, timeout \\ 20000) do
```
Comment (Hajto) [inline test/support/macro/connection_helpers.ex:2]: I don't quite grasp the idea of this function. What exactly does this do? Why not wrap around assert receive if that is really what you need/want?

**Review by Hajto — CHANGES_REQUESTED** [review]

File: test/support/module/fake_transport.ex (hunk 2 above)
Comment (m00lecule) [reply test/support/module/fake_transport.ex]: This module has been deleted 4 days ago

File: test/support/macro/connection_helpers.ex (hunk 13 above)
Comment (m00lecule) [reply test/support/macro/connection_helpers.ex:2]: https://stackoverflow.com/questions/41543466/how-to-assert-elixir-messages-arrive-in-order

File: test/connection_test.exs (hunk 9 above)
Comment (m00lecule) [reply test/connection_test.exs:165]: `assert_called` works only with mocked modules and Mock does it globally. If we want to keep `Janus.ConnectionTest` asynchronous then it is forbidden to mock modules that might be used in other tests.

File: test/session_test.exs (hunk 12 above)
Comment (m00lecule) [reply test/session_test.exs:51]: Thanks, done

File: lib/connection.ex (hunk 8 above)
Comment (m00lecule) [reply lib/connection.ex]: After brief conversation we came to conclusion that tuple argument should be replaced with map (or consider record) and to apply 2.a scenario.

File: test/connection_test.exs (hunk 11 above)
Comment (m00lecule) [reply test/connection_test.exs:1]: `Janus.MockTransport` lacks in configuration options. For example it doesnt allow to mock a situation when `connect/1` returns `{:error, reason}` or respond to `send/1` with other response then `{:ok, _}`. I suggest to make another issue (or attach its change in this PR) to make those changes. If those changes would be applied I could get rid of `BrokenTransport`.

File: test/connection_test.exs (hunk 10 above)
Comment (m00lecule) [reply test/connection_test.exs:30]: Done, but `Mox.stub/3` must be present in `setup` block. Also state that has `table` dependency must stay there.

File: lib/connection.ex [hunk 14]
```diff
@@ -71,7 +72,12 @@ defmodule Janus.Connection do
   defp do_start(method, transport_module, transport_args, handler_module, handler_args, options) do
     apply(GenServer, method, [
       __MODULE__,
-      {transport_module, transport_args, handler_module, handler_args},
+      %{
+        transport_module: transport_module,
```
Comment (m00lecule) [inline lib/connection.ex]: Personally I dont like this change but I could have not found solution to create a map from variable, where its key would be a variable name. Something like `%{transport_module, transport_args ...}`

Comment (Hajto) [reply lib/connection.ex]: Extract this map into a separate variable. It probably will be easier to read.

File: test/connection_test.exs (hunk 9 above)
Comment (Hajto) [reply test/connection_test.exs:165]: Add TODO to replace this with MockTransport when we add function to assert on length of response queue inside the MockTransport.

File: test/support/macro/connection_helpers.ex [hunk 15]
```diff
@@ -18,14 +19,21 @@ defmodule Janus.HandlerTest.CallbackHelper do
     quote do
       test "#{inspect(unquote(fun))} callback" do
         alias Janus.Connection
-        alias Janus.HandlerTest.FakeHandler
-        alias Janus.HandlerTest.FakeHandler.Payloads
+        alias Janus.MockTransport
+        alias Janus.Support.FakeHandler
+        alias FakeHandler.Payloads
```
Comment (Hajto) [inline test/support/macro/connection_helpers.ex]: I would prefer alias by full path.

**Review by Hajto — CHANGES_REQUESTED** [review]

File: test/connection_test.exs (hunk 11 above)
Comment (Hajto) [reply test/connection_test.exs:1]: TBH you can just send a janus response that will trigger error. If you return message that follows schema:
```elixir
     %{
           "janus" => "error",
           "transaction" => transaction,
           "error" => %{"code" => code, "reason" => reason}
         }
```

Then the function call will return an error. I don't know whether this satisfies your needs though.

Comment (m00lecule) [reply test/connection_test.exs:1]: Actually not, just using MockTransport i cant reach all cases:

`case transport_module.send(payload_with_transaction, timeout, transport_state) do
      {:ok, new_transport_state} -> ...
      {:error, reason} -> ...
    end`

and I would be missing `{:error, _}` cause `MockTransport` implements `send/1` this way


  `def send(payload, _timeout, %{pairs: pairs} = state) do
    ...
    {:ok, %{state | pairs: pairs}}
  end`

but it is minor issue that is not worth our time

**Review by Hajto — APPROVED** [review]
